serializer.rst

```diff
@@ -234,195 +454,1556 @@ You can also specify the context on a per-property basis::
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <!-- ... -->
+
+                <!-- if you're using autoconfigure, the tag will be automatically applied -->
+                <service id="Symfony\Component\Serializer\Normalizer\CustomNormalizer">
+                    <!-- register the normalizer with a high priority (called earlier) -->
+                    <tag name="serializer.normalizer"
+                        priority="500"
+                    />
+                </service>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // config/services.php
+        namespace Symfony\Component\DependencyInjection\Loader\Configurator;
+
+        use Symfony\Component\Serializer\Normalizer\CustomNormalizer;
+
+        return function(ContainerConfigurator $container) {
+            // ...
+
+            // if you're using autoconfigure, the tag will be automatically applied
+            $services->set(CustomNormalizer::class)
+                // register the normalizer with a high priority (called earlier)
+                ->tag('serializer.normalizer', [
+                    'priority' => 500,
+                ])
+            ;
+        };
+
+:class:`Symfony\\Component\\Serializer\\Normalizer\\CustomNormalizer`
+    This normalizer calls a method on the PHP object when normalizing. The
+    PHP object must implement :class:`Symfony\\Component\\Serializer\\Normalizer\\NormalizableInterface`
+    and/or :class:`Symfony\\Component\\Serializer\\Normalizer\\DenormalizableInterface`.
+
+:class:`Symfony\\Component\\Serializer\\Normalizer\\GetSetMethodNormalizer`
+    This normalizer is an alternative to the default ``ObjectNormalizer``.
+    It reads the content of the class by calling the "getters" (public
+    methods starting with ``get``, ``is`` or ``has``). It will denormalize
+    data by calling the constructor and the "setters" (public methods
+    starting with ``set``).
+
+    Objects are normalized to a map of property names and values (names are
+    generated by removing the ``get`` prefix from the method name and transforming
+    the first letter to lowercase; e.g. ``getFirstName()`` -> ``firstName``).
+
+:class:`Symfony\\Component\\Serializer\\Normalizer\\PropertyNormalizer`
+    This is yet another alternative to the ``ObjectNormalizer``. This
+    normalizer directly reads and writes public properties as well as
+    **private and protected** properties (from both the class and all of
+    its parent classes) by using `PHP reflection`_. It supports calling the
+    constructor during the denormalization process.
+
+    Objects are normalized to a map of property names to property values.
```

Comment:
So we have 2 older iterations of the ObjectNormalizer still hanging around in the component. Is there any situation where I would *not* want to use the `ObjectNormalizer` and instead register one of these 2? If not, should we deprecate them (or one of them)? Should we stop documenting them, and hope their usages will slowly extinguish?

I really dislike having to say "yet another alternative" without any concrete example of why you should want to use it.

Reply:
Not an expert in this component, but what you say makes sense. Let's keep the good one and deprecate/remove the others.